suite/tests/client-interface/drpttracer_SUDO-test.dll.c

```diff
@@ -0,0 +1,213 @@
+
+    ok = drpttracer_init();
+    CHECK(ok, "drpttracer_init failed");
+
+    dr_register_exit_event(event_exit);
+
+    ok = drmgr_register_thread_init_event(event_thread_init) &&
+        drmgr_register_thread_exit_event(event_thread_exit) &&
+        drmgr_register_pre_syscall_event(event_pre_syscall) &&
+        drmgr_register_post_syscall_event(event_post_syscall);
+    CHECK(ok, "drmgr_register_*_event failed");
+
+    dr_register_filter_syscall_event(event_filter_syscall);
+
+    tls_idx = drmgr_register_tls_field();
+    CHECK(tls_idx > -1, "unable to reserve TLS field");
+}
+
+static void
+event_exit(void)
+{
+    drpttracer_exit();
+    dr_unregister_filter_syscall_event(event_filter_syscall);
+
+    bool ok = drmgr_unregister_thread_init_event(event_thread_init) &&
+        drmgr_unregister_thread_exit_event(event_thread_exit) &&
+        drmgr_unregister_pre_syscall_event(event_pre_syscall) &&
+        drmgr_unregister_post_syscall_event(event_post_syscall) &&
+        drmgr_unregister_tls_field(tls_idx);
+    CHECK(ok, "drmgr_unregister_*_event failed");
+
+    drmgr_exit();
+    dr_fprintf(STDERR, "all done\n");
+}
+
+static void
+event_thread_init(void *drcontext)
+{
+    per_thread_t *pt = (per_thread_t *)dr_thread_alloc(drcontext, sizeof(*pt));
+    memset(pt, 0, sizeof(*pt));
+    drmgr_set_tls_field(drcontext, tls_idx, (void *)pt);
+}
+
+static void
+event_thread_exit(void *drcontext)
+{
+    per_thread_t *pt = (per_thread_t *)drmgr_get_tls_field(drcontext, tls_idx);
+    /* If the thread's last syscall doesn't trigger post_syscall event, we need to end the
+     * tracing manually.
+     */
+    if (pt->tracer_handle != NULL) {
+        end_tracing_and_dump_trace(drcontext);
+        pt->tracer_handle = NULL;
+    }
+    dr_thread_free(drcontext, pt, sizeof(*pt));
+}
+
+static bool
+event_filter_syscall(void *drcontext, int sysnum)
+{
+    return true;
+}
+
+static bool
+event_pre_syscall(void *drcontext, int sysnum)
+{
+    per_thread_t *pt = (per_thread_t *)drmgr_get_tls_field(drcontext, tls_idx);
+    /* If last syscall doesn't trigger post_syscall event, we need to stop its tracing
+     * here.
+     * XXX: We don't stop tracing when the application's system call returns. So we might
+     * trace some system calls called by Dynamorio's internal code.
+     */
+    if (pt->tracer_handle != NULL) {
+        end_tracing_and_dump_trace(drcontext);
+        pt->tracer_handle = NULL;
+    }
+    /* Start trace before syscall. */
+    bool ok = drpttracer_start_tracing(drcontext, DRPTTRACER_TRACING_ONLY_KERNEL, 8, 8,
+                                       &pt->tracer_handle) == DRPTTRACER_SUCCESS;
+    CHECK(ok, "drpttracer_start_tracing failed");
+    return true;
+}
+
+static void
+event_post_syscall(void *drcontext, int sysnum)
+{
+    per_thread_t *pt = (per_thread_t *)drmgr_get_tls_field(drcontext, tls_idx);
+    /* If the syscall doesn't trigger pre_syscall event, we need to skip it. */
```

Comment:
Can it happen that we have a pre-syscall event but not a post-syscall one?

Reply:
Yes. Like the sys_exit.

Reply:
For now, let's add: `ASSERT(pt-:>tracer_handle != NULL)`. I believe currently there's no case where the post_syscall event will find inactive PT tracing. We can revisit if the assert fails anytime.

Reply:
Done.